editoast/src/core/mq_client.rs

```diff
@@ -0,0 +1,236 @@
+            message_type,
+            respond_to,
+            infra_expected_version,
+        }
+    }
+
+    fn serialized_payload(&self) -> Result<Vec<u8>, Error> {
+        Ok(to_vec(&self).map_err(Error::SerializationError)?)
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Lapin(lapin::Error),
+    SerializationError(serde_json::Error),
+    ResponseTimeout,
+}
+
+impl RabbitMQClient {
+    pub async fn new(options: Options) -> Result<Self, lapin::Error> {
+        let connection = Connection::connect(&options.uri, ConnectionProperties::default()).await?;
+        Ok(RabbitMQClient {
+            connection,
+            core_queue_prefix: options.core_queue_prefix,
+            exchange: options.exchange,
+            timeout: options.timeout,
+        })
+    }
+
+    /// Create the core queue.
+    /// If it does not exist, the queue will be created with the provided options.
+    /// If it exists, the queue_declare call will not have any effect.
+    pub async fn create_queue_for_core(
+        &self,
+        channel: &Channel,
+        infra_id: i64,
+    ) -> Result<String, Error> {
+        let core_queue_name = format!("{}-{}", self.core_queue_prefix, infra_id);
+
+        channel
+            .queue_declare(
+                &core_queue_name,
+                QueueDeclareOptions {
+                    exclusive: false,
+                    auto_delete: false,
+                    durable: true,
+                    ..Default::default()
+                },
+                FieldTable::default(),
+            )
+            .await
+            .map_err(Error::Lapin)?;
+
+        Ok(core_queue_name)
+    }
+
+    pub async fn call<T>(
+        &self,
+        message_type: &str,
+        infra_id: i64,
+        infra_expected_version: i64,
+        payload: T,
+    ) -> Result<(), Error>
+    where
+        T: Serialize + Debug,
+    {
+        // Create a channel
+        let channel = self
+            .connection
+            .create_channel()
+            .await
+            .map_err(Error::Lapin)?;
+
+        // Prepare the message
+        let serialized_payload = MessageEnvelope::new(
+            payload,
+            message_type.to_string(),
+            None,
+            infra_expected_version,
+        )
+        .serialized_payload()?;
+
+        // Prepare the queue and publish the message to the core queue
+        let core_queue_name = self.create_queue_for_core(&channel, infra_id).await?;
+        channel
+            .basic_publish(
+                &self.exchange,
+                core_queue_name.as_str(),
+                BasicPublishOptions::default(),
+                &serialized_payload,
+                BasicProperties::default(),
+            )
+            .await
+            .map_err(Error::Lapin)?;
+
+        Ok(())
+    }
+
+    pub async fn call_with_response<T, TR>(
+        &self,
+        message_type: &str,
+        infra_id: i64,
+        infra_expected_version: i64,
+        payload: T,
```

Comment:
why not just take a slice of bytes, return a slice of bytes, and let the caller deal with it?